### PR TITLE
reexport Data.Bifunctor

### PR DESCRIPTION
**Pull request by goolord**
motivation:
consider the following code:
```haskell
decodeFoo :: String -> Either Stirng Foo
decodeFoo = undefined

-- error message doesn't tell the full story
decodeBigFoo :: String -> Either String BigFoo
decodeBigFoo =
      BigFoo
  <$> decodeFoo
  <*> decodeFoo
  <*> decodeFoo

-- a little verbose
decodeBigFooBetter :: String -> Either String BigFoo
decodeBigFooBetter =
      BigFoo
  <$> either Left (Right . ("1: " <>)) decodeFoo
  <*> either Left (Right . ("2: " <>)) decodeFoo
  <*> either Left (Right . ("3: " <>)) decodeFoo

decodeBigFooBest :: String -> Either String BigFoo
decodeBigFooBest =
      BigFoo
  <$> first ("1: " <>) decodeFoo
  <*> first ("2: " <>) decodeFoo
  <*> first ("3: " <>) decodeFoo
```

i also just find `Bifunctor` to be useful generally.

### CODE REVIEW

**Comment by tfausak** [comment]
I'm on the fence about this. I suspect we'll mostly use `Bifunctor` for `Either a b` and `(a, b)`, in which case I would prefer monomorphic functions like `mapLeft :: (a -> c) -> Either a b -> Either c b` or `mapSnd :: (b -> c) -> (a, b) -> (a, c)`. Is the slight increase in convenience worth more than the slight decrease in specificity? Not sure. 

In my mind it's a lot like generalizing `(.)` from functions to categories. Sure `(<<<)` is "more general", but if you only ever use it with one type then you're not really getting anything out of it.

**Comment by goolord** [comment]
i feel like `Bifunctor` is a pretty intuitive abstraction, i think `first, second, bimap` are as powerful but not as common as fmap. whereas Control.Category is not intuitive (requires basic understanding of category thoery), not as powerful (most haskell is written in the category `Hask` (`Set` + `_|_`).

i feel moderately more favorable about `Bifunctor` since it's the abstraction over `Either` that is already in base.

although more broadly, i tend to prefer overloaded versions of functions because it's easier for me to remember the api of an abstraction than it is the api of each type, and i don't feel that that kind of specificity typically buys anything vis a vis type inference or legibility (sometimes it does, not in the case of `Bifunctor` imo)

**Comment by goolord** [comment]
tldr: `Bifunctor` good abstraction, `Control.Category` bad abstraction (which i feel is a more useful metric)

**Comment by saralich** [comment]
I like this idea and I would be fine with adding this but I default to @tfausak's expertise for if we should add it into our Prolude.

**Comment by tfausak** [comment]
I'm fine with it either way. One of the clear benefits of `Bifunctor` is that the functions actually exist. None of `mapLeft`, `mapRight`, `mapFst`, or `mapSnd` are in `base`. 

I'm a little concerned about opening the door for `Control.Arrow`. Both `(&&&)` and `(***)` are sometimes useful but (in my opinion) much harder to understand than their equivalents.

**Comment by goolord** [comment]
i haven't thought enough about the `Category + Arrow` abstractions to have an opinion on how practical of an abstraction it is, but i think it's pretty safe to say using `Control.Arrow` for currying over `(->)` is probably bad, ya

**Comment by tfausak** [comment]
I've seen it used a lot to clean up this pattern:

``` hs
Data.Map.fromList $ fmap (\ x -> (k x, v x)) xs
-- becomes
Data.Map.fromList $ fmap (k &&& v) xs
```

**Review by tfausak — APPROVED** [review]
I'm on the fence here, and these functions are useful, so 👍🏻 from me. Curious to hear from the rest of the @EdutainmentLIVE/engineers though.

**Comment by aschmois** [comment]
I can't see myself using it often but I'd be curious if this could prove me wrong!
